File: go-side/internal/layers/link/link.go
package link

import (
	"fmt"
	"strings"

	"cc3067/lab2/go-side/internal/algorithms"
)

const (
	Hamming = "hamming"
	CRC32   = "crc32"
)

type IntegrityResult struct {
	OK          bool
	MessageBits string
	Corrected   bool
	Error       string
	Syndrome    int
}

func NormalizeAlgorithm(value string) (string, error) {
	normalized := strings.ReplaceAll(strings.ToLower(strings.TrimSpace(value)), "-", "")
	switch normalized {
	case "1", "hamming":
		return Hamming, nil
	case "2", "crc", "crc32":
		return CRC32, nil
	default:
		return "", fmt.Errorf("algoritmo invalido; use hamming o crc32")
	}
}

func CalcularIntegridad(messageBits, algorithm string) (string, error) {
	selected, err := NormalizeAlgorithm(algorithm)
	if err != nil {
		return "", err
	}
	if selected == Hamming {
		return algorithms.HammingEncode(messageBits)
	}
	return algorithms.CRC32Encode(messageBits)
}

func VerificarIntegridad(frameBits, algorithm string, messageLength int) IntegrityResult {
	selected, err := NormalizeAlgorithm(algorithm)
	if err != nil {
		return IntegrityResult{Error: err.Error()}
	}
	if selected == Hamming {
		result := algorithms.HammingDecode(frameBits, messageLength)
		return IntegrityResult{result.Valid, result.DataBits, result.Corrected, result.Error, result.Syndrome}
	}
	ok, data, message := algorithms.CRC32Verify(frameBits, messageLength)
	return IntegrityResult{OK: ok, MessageBits: data, Error: message}
}

func CorregirMensaje(frameBits string, messageLength int) IntegrityResult {
	result := algorithms.HammingDecode(frameBits, messageLength)
	return IntegrityResult{result.Valid, result.DataBits, result.Corrected, result.Error, result.Syndrome}
}

func RedundancyBits(messageLength int, algorithm string) (int, error) {
	selected, err := NormalizeAlgorithm(algorithm)
	if err != nil {
		return 0, err
	}
	if selected == Hamming {
		return algorithms.RequiredParityBits(messageLength)
	}
	return 32, nil
}

File: proyecto/python-side/atm/layers/link.py
"""Capa de Enlace: calculo, verificacion y correccion de integridad."""

from dataclasses import dataclass

from atm.algorithms import crc32, hamming

HAMMING = "hamming"
CRC32 = "crc32"
ALGORITHMS = (HAMMING, CRC32)


@dataclass(frozen=True)
class IntegrityResult:
    """Respuesta uniforme de Enlace para algoritmos de deteccion o correccion."""

    ok: bool
    message_bits: str = ""
    corrected: bool = False
    error: str | None = None
    syndrome: int = 0


def normalize_algorithm(algorithm: str) -> str:
    """Convierte las opciones de consola al identificador canonico del protocolo."""
    value = algorithm.strip().lower().replace("-", "")
    aliases = {"1": HAMMING, "hamming": HAMMING, "2": CRC32, "crc": CRC32, "crc32": CRC32}
    if value not in aliases:
        raise ValueError("algoritmo invalido; use hamming o crc32")
    return aliases[value]


def calcular_integridad(message_bits: str, algorithm: str) -> str:
    """Agrega la redundancia definida por el algoritmo seleccionado."""
    selected = normalize_algorithm(algorithm)
    if selected == HAMMING:
        return hamming.encode(message_bits)
    return crc32.encode(message_bits)


def verificar_integridad(frame_bits: str, algorithm: str, message_length: int) -> IntegrityResult:
    """Verifica una trama; delega en corregir_mensaje cuando el algoritmo puede reparar errores."""
    selected = normalize_algorithm(algorithm)
    if selected == HAMMING:
        return corregir_mensaje(frame_bits, message_length)

    ok, data, error = crc32.verify(frame_bits, message_length)
    return IntegrityResult(ok, data, False, error)


def corregir_mensaje(frame_bits: str, message_length: int) -> IntegrityResult:
    """Servicio de Enlace: verifica Hamming y corrige el bit senalado por el sindrome."""
    result = hamming.decode(frame_bits, message_length)
    return IntegrityResult(result.valid, result.data_bits, result.corrected, result.error, result.syndrome)


def redundancy_bits(message_length: int, algorithm: str) -> int:
    """Calcula el overhead teorico sin construir la trama completa."""
    selected = normalize_algorithm(algorithm)
    return hamming.required_parity_bits(message_length) if selected == HAMMING else 32
